### ISSUE TITLE
Add support for equirectangular panorama

### ISSUE THREAD

**Issue by NerdDoc** [photoprism/photoprism#352]
That will be really cool, if you will add support it.
https://pannellum.org

**Comment by lastzero** [comment]
Can you provide us with free to use pictures for development, demo and testing?

**Comment by NerdDoc** [comment]
> Can you provide us with free to use pictures for development, demo and testing?

Sure. Here is some example photos, and several of them even with gps tags.
![20200524-115521-Unknown-2020-1xf](https://user-images.githubusercontent.com/35555490/83999864-4e185100-a96c-11ea-82b8-82caed5758b1.jpg)
![SAM_100_0395](https://user-images.githubusercontent.com/35555490/83999875-52dd0500-a96c-11ea-996f-e28aef2b2ca9.jpg)
![SAM_100_0496](https://user-images.githubusercontent.com/35555490/83999884-57092280-a96c-11ea-9f7d-8941a73663c8.jpg)
![SAM_100_0509](https://user-images.githubusercontent.com/35555490/83999893-5a9ca980-a96c-11ea-8bfc-1620a31fa25d.jpg)
![SAM_100_0515](https://user-images.githubusercontent.com/35555490/83999902-5d979a00-a96c-11ea-856b-ca57f7a73967.jpg)

**Comment by lastzero** [comment]
Thanks! How did you create them? Are there common apps for this?

**Comment by NerdDoc** [comment]
> Thanks! How did you create them? Are there common apps for this?

These photos was created by Samsung Gear 360 cam. Here is the link to app - https://github.com/ultramango/gear360pano, which can be used for creating this kind of photo, and VLC3 can play 360 video files.

**Comment by NerdDoc** [comment]
As i know - equirectangular panoramic photo - it's a standard.

**Comment by sbaum** [comment]
There are many cameras for creating 360 panorama images. (I'm owner of a Xiaomi Mi Sphere.) Equirectangular projection is the usual way to store these spherical panoramas in a "flat" image format. 
Some basic technical info regarding the respective metadata in equirectangular image files is available e.g. from Facebook: https://facebook360.fb.com/editing-360-photos-injecting-metadata/

**Comment by lastzero** [comment]
Note that we've recently increased the upper limit for JPEGs from 3840 to 7680 pixels (8K Ultra HD). Might be helpful for panoramas too, or are they typically larger?

**Comment by sbaum** [comment]
I think that's sufficient for all current consumer-grade 360° cameras. The Kandao Qoocam 8K outputs photos (as well as videos) in 7680 x 3840 resolution. I don't know any current consumer-grade 360° camera with a higher photo resolution.

**Comment by sbaum** [comment]
Oops, I forgot another method to capture 360° photos: The Google Camera app on Android has a "Photo Sphere" mode where you take a lot of photos all around you. These photos will then automatically be stitched into a equirectangular 360° photo. And these tend to have higher resolutions. I just captured a quick-and-dirty photo-sphere with my Pixel 4 XL smartphone and the resulting image has a resolution of 8704 x 4352 pixels.

**Comment by lastzero** [comment]
Do you typically use originals for 360° without creating smaller, and often sharper, thumbnails? When you create and view them on your personal phone, it obviously doesn't make sense to reduce resolution. Different when you need to download first (or pay for server storage).

**Comment by sbaum** [comment]
Sure, for thumbnails I create smaller files with much lower resolution, and without 360° exif metadata, because it does not make sense to show thumbnails in 360° mode. These thumbnails are just used to find the photo I am looking for. But I would not reduce image resolution when watching a photo in 360° mode. Perhaps I would reduce image size by recompressing the image with lower JPEG quality, if the original file is too big to download. But that's all a question of personal preference. Facebook reduces the resolution of 360° photos after uploading... and as far as I know, most 360° photographers strongly dislike Facebook for doing that, because the lower image quality unfortunately is very noticable.

**Comment by lastzero** [comment]
Photo quality is extremely bad on Facebook in general, but all free hosting services significantly reduce quality to save storage and cpu time.

**Comment by sbaum** [comment]
One additional hint came into my mind: If you are going to generate smaller (=less pixels) versions of 360° photos, you will have to adjust certain exif parameters. Else the smaller versions can not be displayed correctly.

**Comment by lastzero** [comment]
![Screenshot 2020-07-16 at 12 59 38](https://user-images.githubusercontent.com/301686/87663527-41093180-c764-11ea-9d49-54f51aaeaa86.png)
![Screenshot 2020-07-16 at 12 58 51](https://user-images.githubusercontent.com/301686/87663534-42d2f500-c764-11ea-8688-d48c71218596.png)
![Screenshot 2020-07-16 at 12 58 29](https://user-images.githubusercontent.com/301686/87663537-436b8b80-c764-11ea-91b8-4229b13639fc.png)

**Comment by lastzero** [comment]
PhotoPrism now has a panorama filter and indexes the projection type, see 0e5fe5eb43394495df34a7784df03d05adb4211a

Can be tested when our Docker images were updated, either today or tomorrow.

**Comment by NerdDoc** [comment]
COOL!

**Comment by totti4ever** [comment]
Hey, glad to see this is being implemented!

Though, is the feature "only" about filtering/categorizing them or will a proper display be part of?

The categorization/filtering works on my Gear360 and also Google Cardbord Camera files.

**Comment by lastzero** [comment]
We don't have a viewer yet. Panoramas are "just" indexed so that we can easily add this later and you can already use "panorama" as filter.

**Comment by totti4ever** [comment]
Allright, thanks for clarification. I would throw my panos and 360° into Photoprism only when I can display them properly - but I like that you guys are already on the track, please don't get me wrong!

So I'll open a new issue for display purposes, so this ticket with the purpose of handling the panorams in the first place, can be closed, right?

**Comment by lastzero** [comment]
You can leave this open. It's clear that we need a proper panorama viewer before it is done.

I noticed there are different panorama projections though, not just equirectangular... Maybe we'll add follow up issues for those.

**Comment by bleomycin** [comment]
Extremely excited to see support for these is on the radar! I take hundreds of these photos per job in my line of work in extremely high resolution (134 Megapixel per photo) so we can reference tiny details later if needed (construction). Being able to "walk" a project in 360 later is extremely valuable. We currently rely on google photos since it allows proper support with excellent zoom and ipad support but I look forward to the day where I can switch entirely to photoprism.

**Comment by G2G2G2G** [comment]
Never saw this thread, I posted a library here: https://github.com/photoprism/photoprism/issues/1340  
@graciousgrey (not sure if you want to merge these two issues?) 

https://photo-sphere-viewer.js.org/
https://github.com/mistic100/Photo-Sphere-Viewer
> Photo Sphere Viewer can display standard equirectangular panoramas and also cubemaps.

What type of implementation are we looking at? If you classify images as panorama already I guess if we have a boolean like: 
```HTML
<div id="viewer"></div>

<script>
if (imageIsA360Panorama) {
    var viewer = new PhotoSphereViewer.Viewer({
      container: document.querySelector('#viewer'),
      panorama: 'path/to/panorama.jpg'
    });
}
</script>

```

**Comment by graciousgrey** [comment]
@G2G2G2G you are right, being able to display 360degree photos in the photoviewer  (#1340 ) is part of the file type support.

**Comment by bobiko** [comment]
will this functionality appear on production?

**Comment by lastzero** [comment]
Yes, eventually. After we reach our funding goal.

**Comment by Dulanic** [comment]
I'll add my golden label to this, I'd love to finally be rid of 3rd party apps etc... to view my 360/panoramic photos.

Another example of a 360 photo! 

https://dulanic.com/img/PXL_20211230_172250409.PHOTOSPHERE.jpg

**Comment by lastzero** [comment]
Hopefully we have the resources to get this implemented soon, it's not that we don't want to! 🍀

**Comment by kNoAPP** [comment]
I'd financially sponsor this feature being added if the resources required are reasonable. What's needed to finish this out?

**Comment by heikomat** [comment]
Someone recently started integrating photosphere. Is that what you're looking for @kNoAPP ?

https://github.com/fensminger/photoprism/commit/db6b46b6ec0e9ccf41e2f26031b2a86258dcb515

**Comment by kNoAPP** [comment]
Well if @fensminger needs a backer, tester, or a large library of 360 photos (I shot & own them), I'm willing to pledge for the cause! This is exciting stuff.

**Comment by kvalev** [comment]
If someone is up for it, you can cherry pick those two PRs for a fleshed out implementation: https://github.com/kvalev/photoprism/pull/25 & https://github.com/kvalev/photoprism/pull/52

**Comment by ocean2000pl** [comment]
Docker hub image from kvalev/photoprism:preview works on arm (s905, 2gb), but I had to edit manually the type of each photo to "photosphere" via photoprism webui. 
Is it possible to do this automatically? 

When changed manually the type of 5 pictures to "photosphere" within 10 seconds, docker image freezes.

**Comment by kvalev** [comment]
> Docker hub image from kvalev/photoprism:preview works on arm (s905, 2gb), but I had to edit manually the type of each photo to "photosphere" via photoprism webui. Is it possible to do this automatically?

Yes, marking photos as photospheres is done automatically during indexing (using the `IsPhotosphere` EXIF field). If your camera uses a different format, I will have to take a look.  I think if you reindex (maybe you have to do a complete rescan, I am not 100% sure), the rest of the photo sphere images will be detected automatically.

> When changed manually the type of 5 pictures to "photosphere" within 10 seconds, docker image freezes.

Can you be a bit more specific? What do you mean by "freezes"? Feel free to create an issue/discussion [here](https://github.com/kvalev/photoprism) not to pollute this issue.

And one PSA: switching to the `kvalev/photoprism` image will make it a bit hard to switch back to the official release, as there are few incompatible changes.

**Comment by epheterson** [comment]
Excited for this feature! Would be great to also support 360° videos. 

The package noted in the first post supports video, not sure if that was used but if so video may work for free? https://pannellum.org/

**Comment by kNoAPP** [comment]
Looking forward to this feature. I've played with kvalev's fork with the 360 photo support, and its extremely cool/convenient.

**Comment by artincube** [comment]
Looking forward also.
I have thousands of equirectangular stills and videos shot with half dozen different devices (Samsung Gear 360, Ricoh Theta, Insta360 different models, up to 8x4k. I have also shot for work giga 360s of resolution up to several hundred thousand pixels wide 2/1 equirectangular with thousands of frames from a 5DIII on a robot.
I cannot obviously expect Photoprism to manage all these; I even keep many the original files before processing to equirectangular; these show usually two fisheye occupying half of the 2/1 frame, for a two lens/sensor camera, before they are stitched together, and are dng for some cameras. Insta360 have their .insv and .insp formats (exported processed in equirectangular from their software as prores and dng at max settings).
Photoprism way, the fisheye originals could be stacked with the processed versions (stitched into equirectangular), despite have no visual similarity.
Photoprism will be able to display the processed versions, or a conversion from them, in a simple 360 viewer that could provide panning and zooming in and out (change of fov, field of view).
Ideally the user could define global and specific default camera angles and fov on viewer load; these could be saved in the sidecar yml files.
Some basic projections could be also available (little planet...) and setup as defaults.

If you need a selection of files from these cameras let me know.

**Comment by mlapida** [comment]
Also looking forward to this! Happy to contribute additional test images if needed. +1 to being able to stack originals with the pano/sphere version. When shooting with a DJI drone, both the panorama/360 file and the individual JPEGs are saved; it would be great to store them in Photoprism without cluttering up the main search.

**Comment by danielhoherd** [comment]
FYI the DJI drone 360 panorama files that I have looked at do not include the `IsPhotosphere` tag:

```
$ exiftool -s -isPhotosphere -ImageSize -UniqueCameraModel -UsePanoramaViewer -ProjectionType -XPKeywords -- "$fn"
ImageSize                       : 12000x6000
UniqueCameraModel               : DJI FC8482 Mini4 Pro
UsePanoramaViewer               : True
ProjectionType                  : equirectangular
XPKeywords                      : pano
```

**Comment by GlassedSilver** [comment]
> Looking forward also. I have thousands of equirectangular stills and videos shot with half dozen different devices (Samsung Gear 360, Ricoh Theta, Insta360 different models, up to 8x4k. I have also shot for work giga 360s of resolution up to several hundred thousand pixels wide 2/1 equirectangular with thousands of frames from a 5DIII on a robot. I cannot obviously expect Photoprism to manage all these; I even keep many the original files before processing to equirectangular; these show usually two fisheye occupying half of the 2/1 frame, for a two lens/sensor camera, before they are stitched together, and are dng for some cameras. Insta360 have their .insv and .insp formats (exported processed in equirectangular from their software as prores and dng at max settings). Photoprism way, the fisheye originals could be stacked with the processed versions (stitched into equirectangular), despite have no visual similarity. Photoprism will be able to display the processed versions, or a conversion from them, in a simple 360 viewer that could provide panning and zooming in and out (change of fov, field of view). Ideally the user could define global and specific default camera angles and fov on viewer load; these could be saved in the sidecar yml files. Some basic projections could be also available (little planet...) and setup as defaults.
> 
> If you need a selection of files from these cameras let me know.

I came here to figure out if video is being looked into or to at least see acknowledgement that this should be done.

For simplicity since photos have some level of support, maybe make that a separate issue for trackability. I think the image and photo viewers probably can share some code, but video is more complex and its final implementation shouldn't block the completion state of this issue nor should video get overlooked because the issue requesting that **also** in a comment rather than in the formal goal of the issue got closed as completed.

**Comment by kNoAPP** [comment]
Has there been any additional progress on this ticket?

**Comment by GlassedSilver** [comment]
> Has there been any additional progress on this ticket?

Well the roadmap tells it's in the ideas stage and no pull request or assignee is listed.
Given it's been given the priority tag I guess a consideration is at least more likely to happen soon-ish than for other long-standing issues.

What helps as well is that 360° cameras that are rather usable have really come down in price ever since the issue got created increasing the user base that'd profit from this implementation.

**Comment by AllenSikorski** [comment]
As a supporter, do I have the ability to also add a priority tag to this project? The 360 viewer is by far the most important feature to me at the moment. I've got loads of images and video created from the 360 GoPro and the Insta360 that I'd love to be able to nativly view. Synology Photos has a 360 viewer, but since moving from Synology Photos to PhotoPrism, this is by far the biggest feature I require that this project is missing.

**Comment by graciousgrey** [comment]
@vanimox 

Thank you very much for your feedback and support. The golden sponsor labels (now priority) could be added in the past by very early supporters in the higher tiers. We are not adding any new ones at the moment as we need to take care of the existing priorities first. Once those are done, we may look to introduce a feature voting system for members.

**Comment by gamersalpha** [comment]
hello, how to activate 360 video and photo panorama like onprism please ,

**Comment by gamersalpha** [comment]
Hello i send my video 360 but i can have the spherical  view, how activate it ?



but when i lunch a photo no spherical view activated ?

![Uploading image.png…]()


why 360 is not activate ??

**Comment by danielhoherd** [comment]
@gamersalpha the feature you are asking for does not exist yet. This GH issue is requesting the feature.

**Comment by gamersalpha** [comment]
Why I can't found no application to do that :'(

**Comment by AllenSikorski** [comment]
Have there been any updates on this request?

**Comment by halkeye** [comment]
@lastzero I see the help wanted label on this. I tried reading all the comments. Is there any objection to just integrating https://pannellum.org/documentation/examples/simple-example/ if model.panorama? it seems to have an MIT license, so i figure it could just go into package.json? or does there need to be more discussions?

I just want to know whats going on before i look into making a PR

**Comment by lastzero** [comment]
@halkeye Sorry for the delayed response! I was traveling, but I'm now back in the office.

It would be amazing to finally add support for this, e.g. by extending the existing viewer (lightbox) to dynamically load a library like Pannellum and display the image:
- https://github.com/photoprism/photoprism/blob/develop/frontend/src/component/lightbox.vue

It could be implemented similarly to the existing video support (the backend might also need some changes to provide the necessary metadata):

https://github.com/photoprism/photoprism/blob/55575623c5a3ea529c7a3b8be0ae7b3ad6008cd3/frontend/src/component/lightbox.vue#L539-L540

Feel free to reach out to me via [Element/Matrix chat](https://link.photoprism.app/chat) if you need help or have questions! 💬

**Comment by mistic100** [comment]
Hello
If I may allow a little auto-promotion : https://photo-sphere-viewer.js.org/ also supports videos, and is highly customisable.  
Also is is still maintained :-) (which unfortunately is not the case of others solutions, namely pannellum, marzipano or panolens)

**Comment by lastzero** [comment]
@mistic100 Sure, you're welcome! We'd be happy to use it if it can be integrated with our existing [PhotoSwipe 5](https://photoswipe.com/) based viewer?

**Comment by mistic100** [comment]
I never used Photo swipe but if it has events for when an item is displayed it should be easy to integrate.
As suggested above you probably want to load the library only on demand, it is based on ThreeJS so not exactly light...

**Comment by lastzero** [comment]
Sounds good! It would also be great to have more image and video files for development and testing. Are there any good, free resources for downloading sample files in different formats and resolutions?

Users are also welcome to [send samples](https://www.photoprism.app/contact#file-samples) to samples@photoprism.app so that we can host them at https://dl.photoprism.app/samples/ and use them for development. 📬

**Comment by lastzero** [comment]
@mistic100 @halkeye The issue description was updated to include the information shared in the comments, as well as a link to [Photo Sphere Viewer](https://photo-sphere-viewer.js.org/). I also updated the related issue for playing 360-degree videos:

- https://github.com/photoprism/photoprism/issues/4718

**Comment by lastzero** [comment]
We have added the files you submitted to our sample archive:

- https://dl.photoprism.app/samples/Formats/Video/MP4/360deg/
- https://dl.photoprism.app/samples/Formats/Image/JPEG/360deg/
- https://dl.photoprism.app/samples/Brands/DJI/Mini%204%20Pro/

Additional files and file types are welcome!
